sql: TestQueryProgress failed

[(sql).TestQueryProgress failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2084372&tab=buildLog) on [master@4c2da99f879bc2bc3922bc421ac8fee80995b134](https://github.com/cockroachdb/cockroach/commits/4c2da99f879bc2bc3922bc421ac8fee80995b134):

```
=== RUN   TestQueryProgress
    TestQueryProgress: test_log_scope.go:77: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestQueryProgress044936620
    TestQueryProgress: test_log_scope.go:58: use -show-logs to present logs inline
    TestQueryProgress: conn_executor_test.go:569: waiting for query to make progress...
    TestQueryProgress: conn_executor_test.go:529: stalling on scan 6 at /Table/53/1/701/0 and waiting for test to unblock...
    TestQueryProgress: conn_executor_test.go:571: query is now stalled. checking progress...
    TestQueryProgress: conn_executor_test.go:587: 
        	Error Trace:	conn_executor_test.go:587
        	Error:      	Expect "executing (38.40%)" to match "executing \([45]\d\."
        	Test:       	TestQueryProgress
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestQueryProgress044936620
--- FAIL: TestQueryProgress (0.39s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestQueryProgress PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestQueryProgress.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql).TestQueryProgress failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2114638&tab=buildLog) on [master@655f2b999ff6ffe735dd0dd62ea0ecbe92c7dee2](https://github.com/cockroachdb/cockroach/commits/655f2b999ff6ffe735dd0dd62ea0ecbe92c7dee2):

```
=== RUN   TestQueryProgress
    TestQueryProgress: test_log_scope.go:77: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestQueryProgress094924351
    TestQueryProgress: test_log_scope.go:58: use -show-logs to present logs inline
    TestQueryProgress: conn_executor_test.go:569: waiting for query to make progress...
    TestQueryProgress: conn_executor_test.go:529: stalling on scan 6 at /Table/53/1/751/0 and waiting for test to unblock...
    TestQueryProgress: conn_executor_test.go:571: query is now stalled. checking progress...
    TestQueryProgress: conn_executor_test.go:587: 
        	Error Trace:	conn_executor_test.go:587
        	Error:      	Expect "executing" to match "executing \([45]\d\."
        	Test:       	TestQueryProgress
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestQueryProgress094924351
--- FAIL: TestQueryProgress (0.21s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestQueryProgress PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestQueryProgress.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql).TestQueryProgress failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2123605&tab=buildLog) on [master@0570f1a5a38536dddaec69f68a74ad6ce85b827d](https://github.com/cockroachdb/cockroach/commits/0570f1a5a38536dddaec69f68a74ad6ce85b827d):

```
=== RUN   TestQueryProgress
    TestQueryProgress: test_log_scope.go:77: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestQueryProgress247015039
    TestQueryProgress: test_log_scope.go:58: use -show-logs to present logs inline
    TestQueryProgress: conn_executor_test.go:569: waiting for query to make progress...
    TestQueryProgress: conn_executor_test.go:529: stalling on scan 6 at /Table/53/1/751/0 and waiting for test to unblock...
    TestQueryProgress: conn_executor_test.go:571: query is now stalled. checking progress...
    TestQueryProgress: conn_executor_test.go:587: 
        	Error Trace:	conn_executor_test.go:587
        	Error:      	Expect "executing" to match "executing \([45]\d\."
        	Test:       	TestQueryProgress
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestQueryProgress247015039
--- FAIL: TestQueryProgress (0.23s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestQueryProgress PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestQueryProgress.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/conn_executor_test.go
+++ b/pkg/sql/conn_executor_test.go
@@ -30,6 +30,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/security"
 	"github.com/cockroachdb/cockroach/pkg/sql"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/parser"
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
 	"github.com/cockroachdb/cockroach/pkg/sql/row"
@@ -38,7 +39,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/tests"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/ctxgroup"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
@@ -496,8 +496,6 @@ func TestQueryProgress(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	skip.WithIssue(t, 51356)
-
 	const rows, kvBatchSize = 1000, 50
 
 	defer rowexec.TestingSetScannedRowProgressFrequency(rows / 60)()
@@ -550,7 +548,13 @@ func TestQueryProgress(t *testing.T) {
 	db.Exec(t, `CREATE STATISTICS __auto__ FROM t.test`)
 	const query = `SELECT count(*) FROM t.test WHERE x > $1 and x % 2 = 0`
 
-	ctx, cancel := context.WithCancel(context.Background())
+	// Invalidate the stats cache so that we can be sure to get the latest stats.
+	var tableID descpb.ID
+	ctx := context.Background()
+	require.NoError(t, rawDB.QueryRow(`SELECT id FROM system.namespace WHERE name = 'test'`).Scan(&tableID))
+	s.ExecutorConfig().(sql.ExecutorConfig).TableStatsCache.InvalidateTableStats(ctx, tableID)
+
+	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
 	g := ctxgroup.WithContext(ctx)

